misc/drone-scripts/validate-and-serialize.sh

```diff
@@ -13,33 +13,101 @@
-outArray=($prevBuildStatus)
+echo "DRONE_BUILD_NUMBER = $DRONE_BUILD_NUMBER"
 
-# Check if any other build is in the ongoing folder, if present, check if entry is stale.
-# If entry is stale clean it up.
-# If entry is an on going build wait for it.
+# wait time in seconds
+interval=30
 
-prevBuild=`govc datastore.ls $DS docker-volume-vsphere/ongoing|tail -n 1`
+# Decide whether cleanup is needed or not.
+# grab ongoing build# and parse build info, returns true when build_status
+# is "*not* Running" or the request is from the same build number
+function is_cleanup_needed {
+    # Retrieve on-going build information
+    ongoingBuildInfo=`drone build info vmware/docker-volume-vsphere $1`
+    buildInfoObj=($ongoingBuildInfo)
 
-while [[ "$prevBuild" != "" ]];
-do
-    prevBuildStatus=`drone build info vmware/docker-volume-vsphere $prevBuild`
-    outArray=($prevBuildStatus)
-    if [[ ${outArray[2]} != *"running"* || ${outArray[2]} == *"running"* && $prevBuild == $DRONE_BUILD_NUMBER ]]
+    # verifies status value OR request is resulted from *Restart* event
+    if [[ ${buildInfoObj[2]} != *"running"* || $1 == $DRONE_BUILD_NUMBER ]]
     then
-        govc datastore.rm $DS docker-volume-vsphere/ongoing/$prevBuild
+        return 0
     else
-        echo "Waiting 2 minutes for previous build $prevBuild to complete";
-        sleep 120;
+        return 1
+    fi
+}
+
+# Util to remove created directory (consumed as locking mechanism or
+# preserving metadata for ongoing build)
+function release_lock {
+#    echo "Releasing lock file [$1]"
+    govc datastore.rm $DS docker-volume-vsphere/$1
+}
+
+# Util to create directory which in turns consumed as lock
+function acquire_lock {
+    if [ -z "$1" ] ;
+    then
+        echo "Missing argument in acquire_lock" ;
+        echo "USAGE: acquire_lock <lock_name>"
+        return 2;
+    fi
+#    echo "Acquiring the lock  [$1]"
+    govc datastore.mkdir $DS docker-volume-vsphere/$1 2>/dev/null > /dev/null
+    return $?
+}
+
+# First it tries acquire lock "build-start-lock" which makes sure only
+# one thead (test run request gets this opportunity ) and other are
+# being blocked and keeping polling at definite interval.
+# Whichever thread gets success acquiring the "build-start-lock" lock
+# moves forward with kicking off the test run.
+while ! acquire_lock $build_lock
+do
+    isRunAllowed=1
+    # take build check lock
+    if acquire_lock $check_lock
+    then
+        # Grab ongoing build information from drone.
+        runningBuild=`govc datastore.ls $DS docker-volume-vsphere/ongoing`
+        runningBuildArr=($runningBuild)
+        # Let's check cleanup is needed or not
+        # cleans up if return value is 0 otherwise not; checking build is finished or not
+        if is_cleanup_needed ${runningBuildArr[0]}
+        then
+            echo "Cleanning stale data..."
+            release_lock "ongoing/${runningBuildArr[0]}"
+            release_lock $build_lock
+            isRunAllowed=0
```

Comment:
`isRunAllowed` is not needed, as well as lines 97-100 below. instead, just do  (line 92 +) 
```
   datastore.rm <ongoing/id>
   # release_lock $build_lock <= NOT NEEDED AS  YOU NEED TO REACQUIRE IT ANYWAYS
   release_lock $check_lock
   break # this will exit with still holding build lock, which is what you need